searchlib/src/vespa/searchlib/tensor/euclidean_distance.h

```diff
@@ -44,14 +44,19 @@ class SquaredEuclideanDistanceHW : public SquaredEuclideanDistance {
         assert(expected_cell_type() == vespalib::eval::get_cell_type<FloatType>());
     }
 
+    static const double *cast(const double * p) { return p; }
+    static const float *cast(const float * p) { return p; }
+    static const int8_t *cast(const vespalib::eval::Int8Float * p) { return reinterpret_cast<const int8_t *>(p); }
     double calc(const vespalib::eval::TypedCells& lhs, const vespalib::eval::TypedCells& rhs) const override {
         constexpr vespalib::eval::CellType expected = vespalib::eval::get_cell_type<FloatType>();
-        assert(lhs.type == expected && rhs.type == expected);
+        if ((lhs.type != expected) || (rhs.type == expected)) {
```

Comment:
without context: The rhs type should be tested for inequality, not equality. with context: This used to be an invariant that always holds because external code would make sure to cast the query vector to the appropriate (expected) type. Is this true also for int8? if so, is that correct? (we should at least perform a saturated conversion, if at all).

Reply:
I only added this one due to a single test only done for Int8Float
at https://github.com/vespa-engine/vespa/blob/c1f3078567a9553bfb9ba98449937cc8850ccc35/searchlib/src/tests/tensor/distance_functions/distance_functions_test.cpp#L74 where it tests euclidian distance between **Int8Float** tensor and a **double** tensor.

Reply:
Logic flaw with rhs fixed.

Reply:
By switching from the generic euclidean distance function to the HW version (for int8), I think you have forced the query tensors to be converted to int8 rather than float, making values like '200' become undefined (for HNSW).

Reply:
Now we are enforcing input both lhs and rhs to have same and expected type.
Test for mixed double and int8_t has been removed.